localstack/utils/container_utils/container_client.py

```diff
@@ -904,146 +931,193 @@ def untar_to_path(tardata, target_path):
+            action="store_true",
+        )
+        parser.add_argument(
+            "--publish",
+            "-p",
+            help="Publish container port(s) to the host",
+            dest="publish_ports",
+            action="append",
+        )
+        parser.add_argument(
+            "--ulimit", help="Container ulimit settings", dest="ulimits", action="append"
+        )
+        parser.add_argument("--user", "-u", help="Username or UID to execute first process")
+        parser.add_argument(
+            "--volume", "-v", help="Bind mount a volume", dest="volumes", action="append"
+        )
+
+        # Parse
+        flags = shlex.split(additional_flags)
+        args = parser.parse_args(flags)
+
+        # Post-process parsed flags
         extra_hosts = None
-        labels = {}
-        # TODO Use argparse to simplify this logic
-        for flag in shlex.split(additional_flags):
-            if not cur_state:
-                if flag in ["-v", "--volume"]:
-                    cur_state = "volume"
-                elif flag in ["-p", "--publish"]:
-                    cur_state = "port"
-                elif flag in ["-e", "--env"]:
-                    cur_state = "env"
-                elif flag == "--add-host":
-                    cur_state = "add-host"
-                elif flag == "--network":
-                    cur_state = "set-network"
-                elif flag == "--label":
-                    cur_state = "add-label"
-                elif flag in ["-u", "--user"]:
-                    cur_state = "user"
-                elif flag == "--platform":
-                    cur_state = "platform"
-                elif flag == "--privileged":
-                    # Since this is a boolean flag with just one token, the
-                    # override is done here rather than the else block
-                    if privileged is not None:
-                        LOG.warning(
-                            f"Overwriting Docker container privileged flag {privileged} with new value {flag}"
-                        )
-                    privileged = True
-                else:
-                    raise NotImplementedError(
-                        f"Flag {flag} is currently not supported by this Docker client."
-                    )
-            else:
-                if cur_state == "volume":
-                    mounts = mounts if mounts is not None else []
-                    match = re.match(
-                        r"(?P<host>[\w\s\\\/:\-.]+?):(?P<container>[\w\s\/\-.]+)(?::(?P<arg>ro|rw|z|Z))?",
-                        flag,
+        if args.add_hosts:
+            for add_host in args.add_hosts:
+                extra_hosts = extra_hosts if extra_hosts is not None else {}
+                hosts_split = add_host.split(":")
+                extra_hosts[hosts_split[0]] = hosts_split[1]
+
+        if args.envs:
+            for env in args.envs:
+                lhs, _, rhs = env.partition("=")
```

Comment:
it seems we are doing this partitioning in some cases. We could define an argparse custom action for it https://docs.python.org/3/library/argparse.html#action , like demonstrated here: https://www.geeksforgeeks.org/python-key-value-pair-using-argparse/ (was the first link for it but it demonstrates what's possible).

Just a thought, nothing urgent.

Reply:
Fully agree 👍 
I saw it but just tackled the main argument parsing in the first iteration. This looks like a low-hanging fruit though.

Reply:
Just added a comment for now. A quick test revealed that more tests and refinements are needed to ensure equivalent parsing, especially given some flag-specific logic.